runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+			/* Shrink eden a bit */
+			edenRegionChange = edenChangeMagnitude * -1;
+		} else if (_extensions->dnssExpectedTimeRatioMaximum._valueSpecified < hybridEdenOverhead) {
+			/* Expand eden a bit */
+			edenRegionChange = edenChangeMagnitude;
+		}
+
+		_edenSizeFactor += edenRegionChange;	
+	}
+}
+
+double 
+MM_SchedulingDelegate::mapPgcTimeToPgcOverhead(MM_EnvironmentVLHGC *env, uintptr_t partialGcTimeMs) {
+	
+	/* Convert expectedTimeRatioMinimum/Maximum to 0-100 based for this formula */
+	double xminpct = _extensions->dnssExpectedTimeRatioMinimum._valueSpecified * 100;
+	double xmaxpct = _extensions->dnssExpectedTimeRatioMaximum._valueSpecified * 100;
+	double xmaxt = (double)_extensions->tarokTargetMaxPauseTime;
+
+	double overhead;
+
+	if (heapIsFullyExpanded(env)) {
+		/* 
+		 * Eden size is being driven by heuristic which is trying to MINIMIZE hybrid overead. A low avg pgc time, is more desirable than high avg time,
+		 * So the overhead logic needs to map a low avg pgc time, to a low overhead value (aka, a "better"/more desirable value)
+		 * Ex 20ms -> 5% (good/desirable), 1000ms -> 80% (bad/undesirable/eden should probably shrink)
+		 */
+		double midpointPct = (xmaxpct + xminpct)/2.0;
+		if (partialGcTimeMs <= xmaxt) {
+			/* Once the pgc time is at, or below the max pgc time, there is no "benefit" from shrinking it further, since we are already satisfying tarokTargetMaxPauseTime */
+			overhead = midpointPct;
+		} else {
+			/* 
+			 * If pgc time is above the max pgc time, map high PGC time values as very very high overhead, in efforts to bring the PGC time down to tarokTargetMaxPauseTime
+			 * If pgc time is only slightly above tarokTargetMaxPauseTime, then there is only a very small overhead penalty, 
+			 * wheras being 2x higher than the target pause time leads to a significantly bigger penalty  
+			 */
+			double overheadCurve = pow(1.03, ((double)partialGcTimeMs - xmaxt)) + midpointPct - 1;
+			overhead = OMR_MIN(100.0, overheadCurve);
+		}
+		
+	} else {
+		/* 
+		 * Eden sizing logic is trying to keep hybrid overhead between xminpct and xmaxpct, while trying to respect xmaxt. 
+		 * In this situation, when pgc times are very high (above xmaxt), the overhead score needs to return a low number, suggesting contraction.
+		 * If partialGcTimeMs is less than half of xmaxt, eden can expand without any fear of getting close to xmaxt - the mapped cpu overhead here is > xmaxpct (suggesting eden expansion)
+		 * Ex: 20ms -> 12% (suggest expansion), 2000ms -> 0.00% (suggest contraction)
+		 */
+		double slope = (xmaxpct - xminpct) / ((xmaxt/2) - xmaxt);
+		overhead = (slope * (double)partialGcTimeMs) + ((2.0 * xmaxpct) - xminpct);
+		overhead = OMR_MAX(overhead, 0.0);	
+		/*  Expanding simply because pgc time is very small is not a good idea, so return xmaxpct, so that if the pgc cpu overhead wants to expand, only then eden expands */
+		overhead = OMR_MIN(overhead, xmaxpct);
+	}
```

Comment:
I wonder how much contradicting suggestions we might get if heap is close to being fully expanded and sometimes following one path and sometime the other. Ideally, we should not have a hard 'if expanded' check, but have a weighted blend of the 2 recommendations (based on how close to being fully expnded we are), but that would be more pain for future observers to understand. :0

Reply:
Thats a good point. I have observed some small contradictions myself when heap size is fairly close to max size, although what I proposed [in this comment](https://github.com/eclipse/openj9/pull/12043#discussion_r593499193) would be one solution.

I think the solution you suggest (the blend, weighted on how close heap is to being fully expanded), might be another solution worth investigating a bit deeper (unless we choose to take the approach I mention in comment linked just above)

Reply:
Hm, to me those are 2 fairly different problems. If fact with letting overhead going negative you may even amplify the problem I'm trying to address. Very long GC will both affect CPU overhead criteria to try to expand and pause target criteria to try to contract (even more without capping), and with heap being sometimes fully expanded and sometimes just below, will potentially create large oscillations in eden sizing.

Let' give a try to blending....